### ISSUE TITLE
TU `test_analyze_simple_infinite` does not pass on branch `simpl`

### ISSUE THREAD

**Issue by ThomasRuby** [seiller/pymwp#13]
Test on `infinite_2.c` with fc2068cdedf9560879294b71f009ee780cf3ca86

```c
int main(){
    int X0=1, X1=1;
    while(X1<10){
        X0 = X1*X0;
        X1 = X1+X0;
    }
}
```

Fail with :

```python
>           assert str(relation.matrix[0][0].list[2]) == 'i.delta(0,0).delta(0,1)'
E           AssertionError: assert 'i.delta(1,0)' == 'i.delta(0,0).delta(0,1)'
E             - i.delta(0,0).delta(0,1)
E             + i.delta(1,0)
```

Maybe we should check by hand that output values ?

**Comment by nkrusch** [comment]
Yes, we should check by hand. It is possible the test assertion is wrong now after implementing simplification. 

I have another question related to matrix test. I see there is added try catch block for matrix product test here:

https://github.com/seiller/pymwp/blob/fc2068cdedf9560879294b71f009ee780cf3ca86/tests/test_matrix.py#L66

Is that a similar issue that something changes about product because of simplification, or was the try catch added to see error?

**Comment by aubertc** [comment]
> Yes, we should check by hand.

`X0 = X1*X0;` gives

X0 | X1
------------ | -------------
p | o
m | m


X0 | X1
------------ | -------------
m | o
p | m

and

X0 | X1
------------ | -------------
w | o
w | m

`X1 = X1+X0;` gives

X0 | X1
------------ | -------------
m | p 
0 | m


X0 | X1
------------ | -------------
m |  m
o | p

and

X0 | X1
------------ | -------------
m | w
o | w

None of those products will have only m on the diagonal, so there _will be_ infinity on the diagonal, but I'm not sure I understand what this `assert` is saying.

**Comment by seiller** [comment]
@aubertc The assert is about the upper left polynomial only.

@nkrusch I am not sure I understand your question, but we did implement the simplification within the product (which no longer uses the sum since I changed the underlying algorithm, so it was necessary).

**Comment by aubertc** [comment]
> @aubertc The assert is about the upper left polynomial only.

I haven't looked at the code, but the assert may be wrong now: ` 'i.delta(0,0).delta(0,1)'`, hopefully, was simplified: I believe _all the choices_ lead to an `i` in the upper-left corner (if the variables start with `x0` and infinity coefficient are introduced when a coefficient is not `m` on the diagonal but we want to use the Loop rule), so there is no need to make two choices to get an `i`.

Is "all choices give the same thing" represented as "delta(1,0)", with possibly no other "delta(x,0)" in the polynomial?

**Comment by nkrusch** [comment]
@seiller my question was about the reason why `try...catch` was added around the product test?

**Comment by seiller** [comment]
Oh! I do not know. It might be something we overlooked; we played a lot with the assertions to debug. I don't think there is a deep reason for this.

**Comment by nkrusch** [comment]
The unit test that is failing means the following:

After analysis in the resulting matrix, position [0][0] of the matrix contains a polynomial with following monomials (in order):

+m
+i.delta(0,0)
+i.delta(0,0).delta(0,1)
+i.delta(0,0).delta(1,1)

**Comment by seiller** [comment]
@aubertc you may be right. The full list of assertions shows all these monomials should be in the list: 'i.delta(0,0)', 'i.delta(0,0).delta(0,1)', 'i.delta(0,0).delta(1,1)'. In fact, these should reduce to the first one. So the assertions are incorrect, but the question now is: why do we get a third value which has nothing to do with these? (Maybe the assertions did only check the beginning of the list and we are finding coeficients that appeared but after the 4th position; @nkrusch might be able to provide insights on this.)

**Comment by aubertc** [comment]
> In fact, these should reduce to the first one. 

I agree.

> why do we get a third value which has nothing to do with these?

The initial `+m` @nkrusch is mentioning? That I don't know. It may be linked to the way `i` coefficients are introduced, simply added to the original polynomial instead of erasing the previous coefficients?

This connects to my question:
![image](https://user-images.githubusercontent.com/16466689/120698279-d3015200-c47c-11eb-86f8-52227868a20c.png)
Maybe the implementation of L_{\infty} simply _adds_ the monomial with `i` to the original list instead of replacing them?
Still, we may not want to add an `i` it was only an `m` (and not something that actually prevented the application of the loop rule), but, again, this is answered by a bit of the implementation I know little about.

**Comment by nkrusch** [comment]
The meaning of that test was to check \*some deltas\* from the resulting matrix to detect if changes in implementation will cause the analysis results to change. 

In writing that test, I assumed the "analysis produces the correct result and outputs these deltas"

**Comment by seiller** [comment]
@aubertc No the initial 'm' is still here. I am talking about the delta(1,0) that now appears in third position
@nkrusch So you may have tested only the four first monomials in an otherwise longer polynomial? In that case we should just change the assertions and suppose everything works well; all other tests are passed.

**Comment by nkrusch** [comment]
@seiller yes, if the current output is correct, then the action to take is to change the unit test assertions to match current analysis result.

**Comment by seiller** [comment]
Well, I am not sure it is correct.

**Comment by seiller** [comment]
I mean I have no reasons to believe it is incorrect, but I would have liked to understand why the result changed (because I do not have any reason to believe it is correct either).

**Comment by nkrusch** [comment]
Writing that test initially, I also had no "evidence" to say it was correct, I made the assumption it was.

**Comment by seiller** [comment]
Alright. We should change the test then.

**Comment by nkrusch** [comment]
OK, yes, update the assets to match the current outputs

**Comment by nkrusch** [comment]
OK, yes, update the assets to match the current outputs

**Comment by nkrusch** [comment]
OK, yes, update the asserts to match the current outputs

**Comment by seiller** [comment]
done. but it raises the question of further simplification; delta(0,0)+delta(1,0)+delta(2,0) should be one and the whole expression simplifies to i. This is related to what we discussed today as to how we could improve the evaluation running time.

**Comment by ThomasRuby** [comment]
> Oh! I do not know. It might be something we overlooked; we played a lot with the assertions to debug. I don't think there is a deep reason for this.

Yes, it was for debug purpose, in order to print stuff I want before raising the error ^^', I thought could be useful to keep it in the futur…

**Comment by nkrusch** [comment]
@ThomasRuby thanks, usually for unit test, if the test fails we want the error to see that something is not working as expected. If you remove try...catch does the matrix_prod test pass?

**Comment by ThomasRuby** [comment]
> @ThomasRuby thanks, usually for unit test, if the test fails we want the error to see that something is not working as expected. If you remove try...catch does the matrix_prod test pass?

As long as the error is finally `raise`d  at the end of the catch, it will behave the same as without. The only diff (afaik) is that I print more debug info if it fails.
> 
> In a case where test fails, it should show which assertion does not pass, like in the original issue. As such the try catch should not be necessary.

I will show it anyway thanks to keyword `raise`. But you're right, `try … catch` is unnecessary.

**Comment by nkrusch** [comment]
OK leave it there if you find it helpful, I was just curious about the reason

**Comment by ThomasRuby** [comment]
If we agree with @seiller fix 58d4752662559d8fb3a15760ceb3e11051c98564 (which consist in changing assertion) I'm closing this issue.
